internal/pkg/vm/internal/vmcontext/vmcontext.go

```diff
@@ -184,6 +185,122 @@ func (ctx *VMContext) Send(to address.Address, method types.MethodID, value type
 	return out, ret, nil
 }
 
+// Send allows actors to invoke methods on other actors
+func (ctx *VMContext) Send(to address.Address, method types.MethodID, value types.AttoFIL, params []interface{}) interface{} {
+	// check if side-effects are allowed
+	if !ctx.allowSideEffects {
+		runtime.Abort("Calling Send() is not allowed during side-effet lock")
+	}
+
+	deps := ctx.deps
+
+	// the message sender is the `to` actor, so this is what we set as `from` in the new message
+	from := ctx.toAddr
+	fromActor := ctx.to
+
+	vals, err := deps.ToValues(params)
+	if err != nil {
+		runtime.Abort("failed to convert inputs to abi values")
+	}
+
+	paramData, err := deps.EncodeValues(vals)
+	if err != nil {
+		runtime.Abort("encoding params failed")
+	}
+
+	msg := types.NewUnsignedMessage(from, to, 0, value, method, paramData)
+	if msg.From == msg.To {
+		// TODO 3647: handle this
+		runtime.Abortf("unhandled: sending to self (%s)", msg.From)
+	}
+
+	// fetch id address from init actor if necessary
+	toAddr, err := ctx.resolveActorAddress(msg.To)
+	if err != nil {
+		panic(exitcode.ActorNotFound)
+	}
+
+	toActor, err := deps.GetOrCreateActor(context.TODO(), toAddr, func() (*actor.Actor, error) {
+		return &actor.Actor{}, nil
+	})
+	if err != nil {
+		runtime.Abortf("failed to get or create To actor %s", msg.To)
+	}
+	// TODO(fritz) de-dup some of the logic between here and core.Send
+	innerParams := NewContextParams{
+		From:        fromActor,
+		To:          toActor,
+		ToAddr:      toAddr,
+		Message:     msg,
+		OriginMsg:   ctx.originMsg,
+		State:       ctx.state,
+		StorageMap:  ctx.storageMap,
+		GasTracker:  ctx.gasTracker,
+		BlockHeight: ctx.blockHeight,
+		Ancestors:   ctx.ancestors,
+		Actors:      ctx.actors,
+	}
+	innerCtx := NewVMContext(innerParams)
+
+	return deps.Apply(innerCtx)
+}
+
+func apply(ctx *VMContext) interface{} {
```

Comment:
This is mostly a duplicate of `send`. It should either replace it in this PR, one of `apply` or `send` should call the other, or the common code should be extracted into a function.